### ISSUE TITLE
Error: No Policies Found!

### ISSUE THREAD

**Issue by leizerbeam** [ckotzbauer/kyverno-test-action#130]
Hi Christian - thanks for creating tis action - I am having trouble getting it to work.

Would you be able to take a look at my actions code and repo?

https://github.com/leizerbeam/soteria/actions

**Comment by ckotzbauer** [comment]
Hi @leizerbeam,
thanks for using my action. I'm not completely sure what the error is, but you first checkout the repos main and then at the already existing subfolder main again. Maybe the file structure is not the expected one. Could you add a "ls - l" in the kyverno folder?

**Comment by leizerbeam** [comment]
I am not too familiar with GitHub actions code - I was copying your example 3 (policy code in git), what would I put to list the code?

**Comment by ckotzbauer** [comment]
You can just remove the "Checkout kyverno" step when the YAMLs you want to test and the policy-files are in the same repo.
Otherwise add this between the "Checkout kyverno" and "kyverno-test-policy":
```yaml
- name: List files
  run: ls -l kyverno
```

**Comment by leizerbeam** [comment]
<img width="1578" alt="Screen Shot 2022-04-22 at 3 18 09 PM" src="https://user-images.githubusercontent.com/6957385/164803177-c18ad7fc-8a45-4adb-999b-f14877294c79.png">

Here is the screenshot. The folder is basically the root of my project

**Comment by ckotzbauer** [comment]
Yes, there's a nested kyverno folder from the second checkout. This overwrites the policies from this folder. Please try to remove the second checkout

**Comment by leizerbeam** [comment]
progress! however I noticed it may not running the policy tests 
<img width="1347" alt="Screen Shot 2022-04-26 at 3 33 20 PM" src="https://user-images.githubusercontent.com/6957385/165403663-37062054-3915-4ecc-a8bc-ea5acf4eab9b.png">

I have tested the below policies using the Kyverno CLI and the Admission Controller on my K8s Cluster.

`- name: kyverno-test-policy
        # You may pin to the exact commit or the version.
        # uses: ckotzbauer/kyverno-test-action@1c306c7062d34cf7648b831f8a28af7ecb7a13ae
        uses: ckotzbauer/kyverno-test-action@v2
        with:
          # Github token
          token: ${{ github.token }}
          # Enter all application manifests here (Location Profiles, Data Protection Policies, Application Manifests)
          resource-files: |
            kyverno/backup-export-policy.yaml
          # Test that the resource meets all required data protection policies
          policy-files: |
            kyverno/enforce-3-2-1.yaml
            kyverno/enforce-data-protection-by-label.yaml
            kyverno/enforce-minimum-retentions.yaml
            kyverno/enforce-mission-critical-rpo.yaml
`

**Comment by leizerbeam** [comment]
OK I understand that kyverno "test" functionality is not really ideal of the GitHub actions use case - I believe you are passing those resource-files and policy-files. But in order to generate an "Error" output that would be captured by Github actions, it is preferred to use the kyverno 'apply' which will generate the fail message.

The above code will always show 0 pass, 0 failed unless the 'variables.yaml' is included.I did this with mixed results
<img width="435" alt="Screen Shot 2022-04-26 at 4 42 39 PM" src="https://user-images.githubusercontent.com/6957385/165410244-68d1a2ca-484d-42aa-a027-58965cfc7b26.png">
<img width="1003" alt="Screen Shot 2022-04-26 at 4 42 08 PM" src="https://user-images.githubusercontent.com/6957385/165410248-ff83b8ad-6534-404b-960f-217ba730fc1a.png">
<img width="1003" alt="Screen Shot 2022-04-26 at 4 42 08 PM" src="https://user-images.githubusercontent.com/6957385/165410255-b94a1b9a-d9c2-4f7b-8632-d59369ced0f6.png">

**Comment by leizerbeam** [comment]
Hi - an alternative I see here is to use an additional section the required code to define the test result expectations (pass/fail)

results:
  - policy: disallow-empty-ingress-host
    rule: disallow-empty-ingress-host
    resource: ingress-wildcard-host
    result: pass
    kind: Ingress
  - policy: disallow-empty-ingress-host
    rule: disallow-empty-ingress-host
    resource: minimal-ingress
    result: fail
    kind: Ingress

**Comment by ckotzbauer** [comment]
Okay, some notes from me:
> I have tested the below policies using the Kyverno CLI "apply" CLI vs the "test" functionality. Now I realize there is a config file needed here (kyverno-test.yaml) per https://kyverno.io/docs/kyverno-cli/#run-tests-on-a-git-repo

- Yes, this action uses the "test" command. (Can be added to the docs)
- Yes, there's a test-config-file needed for that. This is dynamically generated from this action accoring your resources and the given policies. No need to write this manually.

> Also, you may want remove the 2nd checkout of the repo in your example since it caused my earlier bug.

You refer to the third example. This is meant to be used for the case, when your policies are in another repository. The first two examples do not have this additional checkout. So I won't remove this from the example, as its intended.

> This action needs another section to trigger pass/fail checks - we need the optional "results" block to define the test result expectations (pass/fail) policy tests. If not, then you get 0pass/0fail every time running kyverno test.

The generated test-configs are using the "pass" result as expectation. This action is not useable for complete "unit-tests" of policies, so no "fail" result is ever expected. This action is designed to test YAMLs against a set of policies and all have to pass to pass this action.
Maybe I didn't fully understand your intention, please clarify if needed.

**Comment by leizerbeam** [comment]
Regarding the current behavior, if the "results" block is not specified, the kyverno test command will always pass, regardless if the policies pass or not, this is because the CLI shows a "0 passed, 0 failed" output. if you see my first screenshot, I had an intentional error, and the action did not fail properly.


![](https://user-images.githubusercontent.com/6957385/165403663-37062054-3915-4ecc-a8bc-ea5acf4eab9b.png)


Then I inserted a manual run command for "kyverno test ." (this accepts the "kyverno-test.yaml" file that I had documented with a results block - this is my 2nd screenshot.


![](https://user-images.githubusercontent.com/6957385/165415598-4d7641e6-1e07-4515-ab13-e61d9f21db20.png)

I think its a good unit test action with these few enhancements.  This is just my recommendation based on practicality, but you're the owner so I will let you decide.  Thanks for listening to my feedback.
